akka-remote/src/main/scala/akka/remote/artery/tcp/SSLEngineProvider.scala

```diff
@@ -84,16 +85,33 @@ class ConfigSSLEngineProvider(protected val config: Config, protected val log: M
   val SSLRequireMutualAuthentication: Boolean = config.getBoolean("require-mutual-authentication")
   val HostnameVerification: Boolean = config.getBoolean("hostname-verification")
 
-  private lazy val sslContext: SSLContext = {
-    // log hostname verification warning once
-    if (HostnameVerification)
-      log.debug("TLS/SSL hostname verification is enabled.")
-    else
-      log.info(
-        LogMarker.Security,
-        "TLS/SSL hostname verification is disabled. See Akka reference documentation for more information.")
+  private val SSLContextCacheTime: FiniteDuration = {
+    import akka.util.JavaDurationConverters._
+    config.getDuration("ssl-context-cache-ttl").asScala
+  }
+
+  // log hostname verification warning once
+  if (HostnameVerification)
+    log.debug("TLS/SSL hostname verification is enabled.")
+  else
+    log.info(
+      LogMarker.Security,
+      "TLS/SSL hostname verification is disabled. See Akka reference documentation for more information.")
 
-    constructContext()
+  private def sslContext: SSLContext = getFromCache
+
+  @volatile private var contextRef : Option[CachedSSLContext] = None
+
+  private case class CachedSSLContext(cached: SSLContext, expires: Deadline)
+
+  private def getFromCache: SSLContext = synchronized {
```

Comment:
I'm not sure that I understand why this needs to be `synchronized`. There's no requirement for `SSLContext` to be a singleton, we could just as easily create it from scratch each time without caching with no functional disadvantages. It's just a very small optimisation to cache it. So, without synchronization, the risk is that there will be two `SSLContext`'s in use at the same time, with one of them overwriting the other in the cache. In my opinion, that's a smaller cost to pay than having to introduce blocking here. We could even switch it to use a CAS operation that ensures that if two things do create a new `SSLContext` at the same time, rather than each using their own, the context that wins when being set is the one that gets used by both, while the other is discarded immediately. I'm not sure if there's any advantages to doing that, but it's an option.

Reply:
ok, if there is no major cost of creating several `SSLContext` we should remove the synchronized and just keep the volatile and let the last one overwrite the cached value